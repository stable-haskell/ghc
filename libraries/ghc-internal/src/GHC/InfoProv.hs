{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE UnboxedTuples #-}
{-# LANGUAGE MagicHash #-}
{-# LANGUAGE NoImplicitPrelude #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  GHC.InfoProv
-- Copyright   :  (c) The University of Glasgow 2011
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  ghc-devs@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC Extensions)
--
-- Access to GHC's info-table provenance metadata.
--
-- /The API of this module is unstable and not meant to be consumed by the general public./
-- If you absolutely must depend on it, make sure to use a tight upper
-- bound, e.g., @base < 4.X@ rather than @base < 5@, because the interface can
-- change rapidly without much warning.
--
-- @since 4.18.0.0
-----------------------------------------------------------------------------

module GHC.InfoProv
    ( InfoProv(..)
    , ipLoc
    , whereFrom
      -- * Internals
    , InfoProvEnt
    , ipeProv
    , peekInfoProv
    ) where

import GHC.Base
import GHC.Ptr (nullPtr)
import GHC.InfoProv.Types

-- | Get information about where a value originated from.
-- This information is stored statically in a binary when `-finfo-table-map` is
-- enabled.  The source positions will be greatly improved by also enabled debug
-- information with `-g3`. Finally you can enable `-fdistinct-constructor-tables` to
-- get more precise information about data constructor allocations.
--
-- The information is collect by looking at the info table address of a specific closure and
-- then consulting a specially generated map (by `-finfo-table-map`) to find out where we think
-- the best source position to describe that info table arose from.
--
-- @since 4.16.0.0
whereFrom :: a -> IO (Maybe InfoProv)
whereFrom obj = do
  ipe <- getIPE obj
  -- The primop returns the null pointer in two situations at the moment
  -- 1. The lookup fails for whatever reason
  -- 2. -finfo-table-map is not enabled.
  -- It would be good to distinguish between these two cases somehow.
  if ipe == nullPtr
    then return Nothing
    else do
      infoProv <- peekInfoProv (ipeProv ipe)
      return $ Just infoProv
