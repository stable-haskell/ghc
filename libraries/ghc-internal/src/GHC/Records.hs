{-# LANGUAGE AllowAmbiguousTypes #-}
{-# LANGUAGE ExplicitForAll #-}
{-# LANGUAGE FunctionalDependencies #-}
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE NoImplicitPrelude #-}
{-# LANGUAGE StandaloneKindSignatures #-}
{-# LANGUAGE Trustworthy #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  GHC.Records
-- Copyright   :  (c) Adam Gundry 2015-2023
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  ghc-devs@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC extensions)
--
-- This module defines the 'HasField' class used by the
-- @OverloadedRecordDot@ extension.  See the
-- <https://downloads.haskell.org/ghc/latest/docs/users_guide/exts/hasfield.html
-- user's guide> for more details.
--
-----------------------------------------------------------------------------

module GHC.Records
       ( HasField(..)
       ) where

-- The import of GHC.Base is for build ordering; see Notes in GHC.Base for
-- more info.
import GHC.Base ()

import GHC.Types (TYPE, Constraint)

-- | Constraint representing the fact that the field @x@ belongs to
-- the record type @r@ and has field type @a@.  This will be solved
-- automatically, but manual instances may be provided as well.

--   HasField :: forall {k} {r_rep} {a_rep} . k -> TYPE r_rep -> TYPE a_rep -> Constraint
--   getField :: forall {k} {r_rep} {a_rep} (x::k) (r :: TYPE r_rep) (a :: TYPE a_rep) . HasField x r a => r -> a
-- NB: The {k} means that k is an 'inferred' type variable, and
--     hence not provided in visible type applications.  Thus you
--     say     getField @"foo"
--     not     getField @Symbol @"foo"
type HasField :: forall {k} {r_rep} {a_rep} . k -> TYPE r_rep -> TYPE a_rep -> Constraint
class HasField x r a | x r -> a where
  -- | Selector function to extract the field from the record.
  getField :: r -> a
