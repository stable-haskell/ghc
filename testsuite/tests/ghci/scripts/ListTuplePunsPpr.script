
:set -XUnboxedTuples -XNoListTuplePuns -XDataKinds
import GHC.Tuple (Solo (MkSolo))

:i ()
:i (##)
:i Solo
:t ()
:t (##)
:t (   )
:t (#   #)
:i (,)
:i (#,#)
:t (,)
:t (#,#)
:t (  ,  )
:t (#  ,  #)
type T = (Int, Int)
type S = MkSolo Int
type L = [Int]
:k (Int, Int)
:i T
:i S
:i L
f i (j, k) = i + j + k :: Int
:t f
:t (\ (_, _) -> ())
:t (\ (MkSolo _) -> ())
