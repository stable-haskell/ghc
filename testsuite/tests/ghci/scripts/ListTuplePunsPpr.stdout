type Unit :: *
data Unit = ()
  	-- Defined in ‘GHC.Tuple’
instance Monoid Unit -- Defined in ‘GHC.Internal.Base’
instance Semigroup Unit -- Defined in ‘GHC.Internal.Base’
instance Bounded Unit -- Defined in ‘GHC.Internal.Enum’
instance Read Unit -- Defined in ‘GHC.Internal.Read’
instance Enum Unit -- Defined in ‘GHC.Internal.Enum’
instance Show Unit -- Defined in ‘GHC.Internal.Show’
instance Eq Unit -- Defined in ‘GHC.Classes’
instance Ord Unit -- Defined in ‘GHC.Classes’
type Unit# :: GHC.Types.ZeroBitType
data Unit# = (##)
  	-- Defined in ‘GHC.Types’
type Solo :: * -> *
data Solo a = MkSolo a
  	-- Defined in ‘GHC.Tuple’
instance Traversable Solo
  -- Defined in ‘GHC.Internal.Data.Traversable’
instance Applicative Solo -- Defined in ‘GHC.Internal.Base’
instance Foldable Solo -- Defined in ‘GHC.Internal.Data.Foldable’
instance Functor Solo -- Defined in ‘GHC.Internal.Base’
instance Monad Solo -- Defined in ‘GHC.Internal.Base’
instance Monoid a => Monoid (Solo a)
  -- Defined in ‘GHC.Internal.Base’
instance Semigroup a => Semigroup (Solo a)
  -- Defined in ‘GHC.Internal.Base’
instance Bounded a => Bounded (Solo a)
  -- Defined in ‘GHC.Internal.Enum’
instance Enum a => Enum (Solo a) -- Defined in ‘GHC.Internal.Enum’
instance Read a => Read (Solo a) -- Defined in ‘GHC.Internal.Read’
instance Eq a => Eq (Solo a) -- Defined in ‘GHC.Classes’
instance Ord a => Ord (Solo a) -- Defined in ‘GHC.Classes’
instance Show a => Show (Solo a) -- Defined in ‘GHC.Internal.Show’
() :: Unit
(##) :: Unit#
(   ) :: Unit
(#   #) :: Unit#
type Tuple2 :: * -> * -> *
data Tuple2 a b = (,) a b
  	-- Defined in ‘GHC.Tuple’
instance Traversable (Tuple2 a)
  -- Defined in ‘GHC.Internal.Data.Traversable’
instance Foldable (Tuple2 a)
  -- Defined in ‘GHC.Internal.Data.Foldable’
instance Monoid a => Applicative (Tuple2 a)
  -- Defined in ‘GHC.Internal.Base’
instance Functor (Tuple2 a) -- Defined in ‘GHC.Internal.Base’
instance Monoid a => Monad (Tuple2 a)
  -- Defined in ‘GHC.Internal.Base’
instance (Monoid a, Monoid b) => Monoid (Tuple2 a b)
  -- Defined in ‘GHC.Internal.Base’
instance (Semigroup a, Semigroup b) => Semigroup (Tuple2 a b)
  -- Defined in ‘GHC.Internal.Base’
instance (Bounded a, Bounded b) => Bounded (Tuple2 a b)
  -- Defined in ‘GHC.Internal.Enum’
instance (Ord a, Ord b) => Ord (Tuple2 a b)
  -- Defined in ‘GHC.Classes’
instance (Read a, Read b) => Read (Tuple2 a b)
  -- Defined in ‘GHC.Internal.Read’
instance (Show a, Show b) => Show (Tuple2 a b)
  -- Defined in ‘GHC.Internal.Show’
instance (Eq a, Eq b) => Eq (Tuple2 a b)
  -- Defined in ‘GHC.Classes’
type Tuple2# :: *
                -> *
                -> TYPE
                     (GHC.Types.TupleRep [GHC.Types.LiftedRep, GHC.Types.LiftedRep])
data Tuple2# a b = (#,#) a b
  	-- Defined in ‘GHC.Types’
(,) :: a -> b -> Tuple2 a b
(#,#) :: a -> b -> Tuple2# a b
(  ,  ) :: a -> b -> Tuple2 a b
(#  ,  #) :: a -> b -> Tuple2# a b
(Int, Int) :: Tuple2 (*) (*)
type T :: Tuple2 (*) (*)
type T = (Int, Int) :: Tuple2 (*) (*)
  	-- Defined at <interactive>:18:1
type S :: Solo (*)
type S = MkSolo Int :: Solo (*)
  	-- Defined at <interactive>:19:1
type L :: List (*)
type L = [Int] :: List (*)
  	-- Defined at <interactive>:20:1
f :: Int -> Tuple2 Int Int -> Int
(\ (_, _) -> ()) :: Tuple2 a b -> Unit
(\ (MkSolo _) -> ()) :: Solo a -> Unit
